resources/bolt_bash_completion.sh

```diff
@@ -6,134 +6,202 @@
+  local run_context_opts="-c --concurrency -i --inventoryfile --save-rerun --no-save-rerun --cleanup --no-cleanup"
+  local transport_opts="--transport --connect-timeout --tty --no-tty --native-ssh --no-native-ssh --ssh-command --copy-command"
+  local display_opts="--format --color --no-color -v --verbose --no-verbose --trace --stream"
+  local action_opts="${inventory_opts} ${context_opts} ${authentication_opts} ${escalation_opts} ${run_context_opts} ${transport_opts} ${display_opts}"
+  local apply_opts="--compile-concurrency --hiera-config"
+  local file_opts=" -m --modulepath --project --private-key -i --inventoryfile --hiera-config "
 
-	local inventory_list_cache_file="/tmp/bolt_inventory_cache_list.$$.tmp"
+  # If there's only one word and it's "bolt", tab complete the subcommand
+  if [ $COMP_CWORD -eq 1 ]; then
+    next="apply command file group guide inventory lookup module plan project script secret task"
+  fi
 
-	if ([[ $prev == -* ]] || [[ $prevprev == -* ]]) && [[ " -t --targets --format --log-level --rerun " != *" ${prev} "* ]]
-	then
-		if [[ ${COMP_WORDS[1]} == "apply" ]]
-		then
-			prev=${COMP_WORDS[1]}
-			prevprev=${COMP_WORDS[0]}
-		else
-			prev=${COMP_WORDS[2]}
-			prevprev=${COMP_WORDS[1]}
-		fi
-	fi
+  # Tab complete files for options that accept files. The spaces are important!
+  # They make it so `bolt inventory` isn't confused with `--inventoryfile`.
+  if [[ $file_opts =~ " ${prev} " ]]; then
+    next=$(compgen -f -d "" -- $cur)
+  # Handle tab completing enumerable CLI options
+  elif [ "$prev" == "--log-level" ]; then
+    next="trace debug info warn error fatal any"
+  elif [ "$prev" == "--transport" ]; then
+    next="docker local lxd pcp podman remote ssh winrm"
+  elif [ "$prev" == "--format" ]; then
+    next="human json rainbow"
+  else
+    # Once we have subcommands, tab complete actions
+    case $subcommand in
+      apply)
+        next="${action_opts} ${apply_opts} -e --execute"
+        ;;
+      command)
+        if [ $COMP_CWORD -eq 2 ]; then
+          next="run"
+        elif [ $action = "run" ]; then
+          next="${action_opts} --env-var"
+        fi
+        ;;
+      file)
+        if [ $COMP_CWORD -eq 2 ]; then
+          next="download upload"
+        else
+          case $action in
+            download) next="${action_opts} --tmpdir";;
+            upload) next=$action_opts;;
+          esac
+        fi
+        ;;
+      group)
+        if [ $COMP_CWORD -eq 2 ]; then
+          next="show"
+        elif [ $action == 'show' ]; then
+          next="--project --format --inventoryfile"
+        fi
+        ;;
+      guide)
+        next="--format"
+        ;;
+      inventory)
+        if [ $COMP_CWORD -eq 2 ]; then
+          next="show"
+        elif [ $action == 'show' ]; then
+          next="${inventory_opts} --project --format --inventoryfile --detail"
+        fi
+        ;;
+      lookup)
+        next="${action_opts} --hiera-config"
```

Comment:
```suggestion
        next="${action_opts} --hiera-config --plan-hierarchy"
```

Now that https://github.com/puppetlabs/bolt/pull/2859 is merged.